// cubing/dist/lib/cubing/chunks/chunk-XO3JAA3V.js
// src/cubing/vendor/apache/comlink-everywhere/node-adapter.ts
function nodeEndpoint(parentPort) {
  const listeners = /* @__PURE__ */ new WeakMap();
  const parentPortAsNodeWorker = parentPort;
  const constructedWorker = {
    postMessage: parentPort.postMessage.bind(parentPort),
    addEventListener: (_, eh) => {
      const l = (data) => {
        if ("handleEvent" in eh) {
          eh.handleEvent({ data });
        } else {
          eh({ data });
        }
      };
      parentPort.on("message", l);
      listeners.set(eh, l);
    },
    removeEventListener: (_, eh) => {
      const l = listeners.get(eh);
      if (!l) {
        return;
      }
      parentPort.off("message", l);
      listeners.delete(eh);
    },
    nodeWorker: parentPortAsNodeWorker,
    terminate: () => {
      parentPortAsNodeWorker.terminate();
    }
    // start: nep.start && nep.start.bind(nep),
  };
  return constructedWorker;
}
var node_adapter_default = nodeEndpoint;

export {
  node_adapter_default
};
